File: ee/codegen/src/utils/vellum-variables.ts
import { python } from "@fern-api/python-ast";
import * as Vellum from "vellum-ai/api";

import { VELLUM_CLIENT_MODULE_PATH } from "src/constants";
import { WorkflowContext } from "src/context";
import { assertUnreachable } from "src/utils/typing";

export function getVellumVariablePrimitiveType({
  type,
  workflowContext,
}: {
  type: Vellum.VellumVariableType;
  workflowContext: WorkflowContext;
}): python.Type {
  switch (type) {
    case "STRING":
      return python.Type.str();
    case "NUMBER":
      return python.Type.float();
    case "JSON":
      return python.Type.reference(
        python.reference({
          name: "Json",
          modulePath: [
            ...workflowContext.sdkModulePathNames.WORKFLOWS_MODULE_PATH,
            "types",
          ],
        })
      );
    case "CHAT_HISTORY":
      return python.Type.list(
        python.Type.reference(
          python.reference({
            name: "ChatMessage",
            modulePath: VELLUM_CLIENT_MODULE_PATH,
          })
        )
      );
    case "SEARCH_RESULTS":
      return python.Type.list(
        python.Type.reference(
          python.reference({
            name: "SearchResult",
            modulePath: VELLUM_CLIENT_MODULE_PATH,
          })
        )
      );
    case "ERROR":
      return python.Type.reference(
        python.reference({
          name: "VellumError",
          modulePath: VELLUM_CLIENT_MODULE_PATH,
        })
      );
    case "ARRAY":
      return python.Type.list(
        python.Type.reference(
          python.reference({
            name: "VellumValue",
            modulePath: VELLUM_CLIENT_MODULE_PATH,
          })
        )
      );
    case "FUNCTION_CALL":
      return python.Type.reference(
        python.reference({
          name: "FunctionCall",
          modulePath: VELLUM_CLIENT_MODULE_PATH,
        })
      );
    case "IMAGE":
      return python.Type.reference(
        python.reference({
          name: "VellumImage",
          modulePath: VELLUM_CLIENT_MODULE_PATH,
        })
      );
    case "AUDIO":
      return python.Type.reference(
        python.reference({
          name: "VellumAudio",
          modulePath: VELLUM_CLIENT_MODULE_PATH,
        })
      );
    case "NULL":
      return python.Type.none();
    default: {
      assertUnreachable(vellumVariableType);
    }
  }
}

File: src/vellum/workflows/utils/vellum_variables.py
import typing
from typing import List, Tuple, Type, Union, get_args, get_origin

from vellum import (
    ChatMessage,
    ChatMessageRequest,
    FunctionCall,
    FunctionCallRequest,
    SearchResult,
    SearchResultRequest,
    VellumAudio,
    VellumAudioRequest,
    VellumError,
    VellumErrorRequest,
    VellumImage,
    VellumImageRequest,
    VellumValue,
    VellumValueRequest,
    VellumVariableType,
)
from vellum.workflows.descriptors.base import BaseDescriptor
from vellum.workflows.types.core import Json


def primitive_type_to_vellum_variable_type(type_: Union[Type, BaseDescriptor]) -> VellumVariableType:
    """Converts a python primitive to a VellumVariableType"""
    if isinstance(type_, BaseDescriptor):
        # Ignore None because those just make types optional
        types = [t for t in type_.types if t is not type(None)]

        # default to JSON for typevars where the types is empty tuple
        if len(types) == 0:
            return "JSON"

        if len(types) != 1:
            # Check explicitly for our internal JSON type.
            # Matches the type found at vellum.workflows.utils.vellum_variables.Json
            if types == [
                bool,
                int,
                float,
                str,
                typing.List[typing.ForwardRef("Json")],  # type: ignore [misc]
                typing.Dict[str, typing.ForwardRef("Json")],  # type: ignore [misc]
            ]:
                return "JSON"
            raise ValueError(f"Expected Descriptor to only have one type, got {types}")

        type_ = type_.types[0]

    if _is_type_optionally_equal(type_, str):
        return "STRING"
    elif _is_type_optionally_in(type_, (int, float)):
        return "NUMBER"
    elif _is_type_optionally_in(type_, (FunctionCall, FunctionCallRequest)):
        return "FUNCTION_CALL"
    elif _is_type_optionally_in(type_, (VellumImage, VellumImageRequest)):
        return "IMAGE"
    elif _is_type_optionally_in(type_, (VellumAudio, VellumAudioRequest)):
        return "AUDIO"
    elif _is_type_optionally_in(type_, (VellumError, VellumErrorRequest)):
        return "ERROR"
    elif _is_type_optionally_in(type_, (List[ChatMessage], List[ChatMessageRequest])):
        return "CHAT_HISTORY"
    elif _is_type_optionally_in(type_, (List[SearchResult], List[SearchResultRequest])):
        return "SEARCH_RESULTS"
    elif _is_type_optionally_in(type_, (List[VellumValue], List[VellumValueRequest])):
        return "ARRAY"

    return "JSON"


def _is_type_optionally_equal(type_: Type, target_type: Type) -> bool:
    if type_ == target_type:
        return True

    origin = get_origin(type_)
    if origin is not Union:
        return False

    args = get_args(type_)
    if len(args) != 2:
        return False

    source_type, none_type = args
    if none_type is not type(None):
        return False

    return _is_type_optionally_equal(source_type, target_type)


def _is_type_optionally_in(type_: Type, target_types: Tuple[Type, ...]) -> bool:
    return any(_is_type_optionally_equal(type_, target_type) for target_type in target_types)
